Sources/MapboxDirections/MapMatching/MatchResponse.swift

```diff
@@ -17,12 +17,14 @@ class MatchResponse: Codable {
         let container = try decoder.container(keyedBy: CodingKeys.self)
         code = try container.decode(String.self, forKey: .code)
         message = try container.decodeIfPresent(String.self, forKey: .message)
-        tracepoints = try container.decodeIfPresent([Tracepoint?].self, forKey: .tracepoints)
+        tracepoints = try container.decodeIfPresent([Match.Tracepoint?].self, forKey: .tracepoints)
         matches = try container.decodeIfPresent([Match].self, forKey: .matches)
         
-        if let points = self.tracepoints {
-            matches?.forEach {
-                $0.tracepoints = points
+        if let tracepoints = self.tracepoints, let matches = matches {
+            for match in matches {
+                // TODO: Filter on matchings_index.
+                // TODO: Push tracepoints down to individual legs to reflect waypoint_index.
```

Comment:
Can someone point me to some kind of a doc or explanation about `tracepoints` in the response and `matchings_index`, `waypoint_index` too?

Reply:
Here’s the [Map Matching API documentation](https://docs.mapbox.com/api/navigation/#map-matching), which is more relevant than before, now that this PR hews more closely to the Map Matching API response format.

It isn’t necessary to fix #386 as part of this PR, but I think understanding that issue helped me understand these properties better.

Reply:
According to that doc, our `Match` and `Tracepoint` objects do not reflect the structure described. Is there any special reason why? Or that's what I can safely take care of?

Reply:
The Map Matching API response format avoids repetition to minimize data transfer, and it’s optimized for a file format (JSON) that has no concept of pointer references. As a result, the response is rife with parallel structures and indices. At the same time, the Directions API response is influenced by the OSRM Router service’s heavily nested response format (maneuvers inside steps inside legs inside routes). As part of developing the “bring your own route” workflow, the Map Matching API adopted most of the Directions API’s format so that the client could duck-type matches as routes. So Map Matching API responses end up with a mix of indexable parallel structures in some places and repetitive, nested structures in other places.

Originally, this library only supported the Directions API, so it adopted that API’s highly nested structure. We also avoided index-based structures, even where the Directions API had them, because of potential out-of-bounds issues and other misuse. Pointer references are more convenient to work with than indices and don’t result in excessive memory consumption. It’s also much more convenient for objects to be self-contained: if the application only cares about the match, it can pass around or archive an individual Match object instead of the entire MatchResponse plus a match index.

In the long term, I think we will want to go all-in on indexed parallel structures, once a future version of the Directions API adopts Valhalla’s more consistently indexed-based structure. When we do that, things like the `RouteLeg.segmentRangesByStep` property added in mapbox/mapbox-directions-swift#250 will be important for avoiding out-of-bounds issues.

But in the short term, we should keep pushing things like `tracepoints` down into the individual matches so that the matches are more self-contained.

This is a good example of how I think we want to be deliberate about where this library’s types resemble or differ from the API response. The API response format has a lot of warts, having bent over backwards to preserve backwards compatibility, so we only want to copy what make sense. We prioritized refactoring the waypoint/tracepoint types because they were too different from the corresponding API response objects, leading to some questionable patterns like treating `RouteOptions.waypoints` like an inout parameter. But I tried to avoid churning other classes were we differed in more minor or more deliberate ways.

Reply:
OK, so due to some reasonable matters like API structure and JSON format we may not follow the doc strictly but instead "adjust" info we provide to users to have some adequate form. 
At the same time, to correctly "keep pushing things like tracepoints down into the individual matches" we need to follow the doc and decode `Tracepoint`'s `matchings_index` and `waypoint_index`... I see that previous implementation just copied all tracepoints to all matches, which seem to work only while there is a single `Match` provided (unless it was removed [here](https://github.com/mapbox/mapbox-directions-swift/commit/48abd96069cbe131f076040b0e1bb3c04ef48bf8#diff-c28392baf0208871bf1ac5c51e283fe1L50-L54)).
We may just restore this behavior, but why not improve it to make it "more safe".